driver-core/src/test/java/com/datastax/driver/core/TabletsIT.java

```diff
@@ -1,113 +1,388 @@
-        Assert.fail("Received non empty payload with tablets routing information: " + payload);
+        throw new RuntimeException(
+            "Received non empty payload with tablets routing information: " + payload);
+      }
+    }
+  }
+
+  private static boolean waitSessionLearnedTabletInfo(Session session) {
+    if (isSessionLearnedTabletInfo(session)) {
+      return true;
+    }
+    // Wait till tablet update, which is async, is completed
+    try {
+      Thread.sleep(200);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+    }
+    return isSessionLearnedTabletInfo(session);
+  }
+
+  private static boolean checkIfRoutedProperly(Session session, Statement stmt) {
+    // DefaultLoadBalancingPolicy suppose to prioritize nodes from replica list randomly shuffling
+    // them
+    // If routing goes wrong you will see more than REPLICATION_FACTOR unique first nodes in
+    // execution plan
+
+    int expectedNodesCount = stmt.isLWT() ? 1 : REPLICATION_FACTOR;
+    Set<Host> nodes = new HashSet<>();
+    for (int i = 0; i < REPLICATION_FACTOR * 3; i++) {
+      nodes.add(session.execute(stmt).getExecutionInfo().getQueriedHost());
+    }
+    return nodes.size() <= expectedNodesCount;
+  }
+
+  private static boolean isSessionLearnedTabletInfo(Session session) {
+    Map<TabletMap.KeyspaceTableNamePair, TabletMap.TabletSet> tabletMapping =
+        session.getCluster().getMetadata().getTabletMap().getMapping();
+    TabletMap.KeyspaceTableNamePair ktPair =
+        new TabletMap.KeyspaceTableNamePair(KEYSPACE_NAME.toLowerCase(), TABLE_NAME.toLowerCase());
+
+    TabletMap.TabletSet tablets = tabletMapping.get(ktPair);
+    if (tablets == null || tablets.tablets.isEmpty()) {
+      return false;
+    }
+
+    for (TabletMap.Tablet tab : tablets.tablets) {
+      if (tab.getReplicas().size() >= REPLICATION_FACTOR) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private static void assertSessionTabletMapIsFilled(Session session) {
+    Map<TabletMap.KeyspaceTableNamePair, TabletMap.TabletSet> tabletMapping =
+        session.getCluster().getMetadata().getTabletMap().getMapping();
+    TabletMap.KeyspaceTableNamePair ktPair =
+        new TabletMap.KeyspaceTableNamePair(KEYSPACE_NAME.toLowerCase(), TABLE_NAME.toLowerCase());
+    Assert.assertTrue(tabletMapping.containsKey(ktPair));
+
+    TabletMap.TabletSet tablets = tabletMapping.get(ktPair);
+    Assert.assertEquals(tablets.tablets.size(), INITIAL_TABLETS);
+
+    for (TabletMap.Tablet tab : tablets.tablets) {
+      Assert.assertEquals(tab.getReplicas().size(), REPLICATION_FACTOR);
+    }
+  }
+
+  private static boolean executeOnAllHostsAndReturnIfResultHasTabletsInfo(
+      Session session, Statement stmt) {
+    for (Host node : session.getCluster().getMetadata().getAllHosts()) {
+      if (executeAndReturnIfResultHasTabletsInfo(session, stmt.setHost(node))) {
+        return true;
       }
     }
+    return false;
+  }
```

Comment:
is it supposed to execute on ALL hosts or on ANY host? Because it will return on first success.

Reply:
It supposed to return false after going over all hosts, and return true after first success